cluster/src/dunit/scala/org/apache/spark/sql/ColumnBatchAndExternalTableDUnitTest.scala

```diff
@@ -278,6 +286,118 @@ class ColumnBatchScanDUnitTest(s: String) extends ClusterManagerTestBase(s) {
     snc.sql("DROP TABLE IF EXISTS " + rowTable)
     snc.sql("DROP TABLE IF EXISTS " + colTable)
   }
+
+  def testSessionConfigForCPUsPerTask(): Unit = {
+    val netPort = AvailablePortHelper.getRandomAvailableTCPPort
+    vm2.invoke(classOf[ClusterManagerTestBase], "startNetServer", netPort)
+    val conn = getANetConnection(netPort)
+    val stmt = conn.createStatement()
+
+    val sparkCores = TestUtils.defaultCores * 3 // three executors
+    val usableHeap = vm1.invoke(new SerializableCallable[AnyRef] {
+      override def call(): AnyRef = {
+        Long.box(Misc.getGemFireCache.getResourceManager.getHeapMonitor
+            .getThresholds.getCriticalThresholdBytes)
+      }
+    }).asInstanceOf[java.lang.Long].longValue()
+    val implicitCpusToTasks = math.max(1, math.ceil(math.max(
+      128.0 * 1024.0 * 1024.0 * TestUtils.defaultCores / usableHeap,
+      TestUtils.defaultCores.toDouble / Runtime.getRuntime.availableProcessors())).toInt)
+    val buckets = math.max(128, sparkCores)
+    val targetDataFile = "airlineParquet"
+
+    // initialize data
+    val srcDataFile = getClass.getResource("/2015-trimmed.parquet").getPath
+    stmt.execute(s"create table airline_staging using parquet options (path '$srcDataFile')")
+    stmt.execute(s"create table airline using column options (buckets '$buckets') " +
+        "as select * from airline_staging")
+    stmt.execute("drop table airline_staging")
+    // create parquet file with large number of partitions though total size is small
+    stmt.execute(s"create table airline_staging using parquet options (path '$targetDataFile') " +
+        s"as select * from airline")
+
+    var maxTasksStarted = 0
+    var activeTasks = 0
+
+    val listener = new SparkListener {
+      override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = synchronized {
+        if (taskStart.taskInfo ne null) {
+          activeTasks += 1
+          if (activeTasks > maxTasksStarted) {
+            maxTasksStarted = activeTasks
+          }
+        }
+      }
+
+      override def onTaskEnd(taskEnd: SparkListenerTaskEnd): Unit = synchronized {
+        if ((taskEnd.taskInfo ne null) && taskEnd.stageAttemptId != -1) {
+          activeTasks -= 1
+        }
+      }
+    }
+    sc.listenerBus.addListener(listener)
+
+    // ---- Check explicit spark.task.cpus setting takes effect in embedded mode -----
+
+    // force linking of buckets to partitions to avoid de-linking complications
+    stmt.execute("set snappydata.linkPartitionsToBuckets = true")
+    // first confirm that number of tasks created at a time are same as sparkCores by default
+    maxTasksStarted = 0
+    activeTasks = 0
+    stmt.execute("select avg(depDelay) from airline")
+    // should not deviate much though in rare cases few tasks can start/end slightly out of order
+    assert(maxTasksStarted < sparkCores * 2)
```

Comment:
TestUtils.defaultCores will be 8*2 so sparkCores will be three times which is 48. Total tasks will be 128 after linking the same as number of buckets.

Regardless maxTasksStarted notes the maximum concurrent number of tasks which should be around 48 though in some cases found that the listener stats are slightly out of order perhaps because they are done after the fact outside of the original sync block, so they could go to around 50 or so. This is a conservative upper-limit.

Reply:
yes instead of "test will calculate sparkcores = 128,"
I meant "test will calculate sparkcores = 48,"

Ok. I see.